### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by limux82** [Notalib/nativescript-accessibility-ext#37]
With the version 6.0.0.alpha2 we experienced a lot of performance problems in iOS when we import the NotaAccessibilityExtModule. It take 6-8 second to made route change and it takes seconds to draw layouts that otherwise are immediate.

Are you working in order to address this problem?

**Comment by m-abs** [comment]
Hi @limux82 

Thank you very much for your feedback.

Our app is still on nativescript 5.4.x, so we haven't run into these issues yet, but of cause they have to be addressed.

I've just published `v6.1.0-alpha.1` with support for theme-2.0 (including compat-mode) and it should give some performance improvements.
It uses the new `css-calc` and `css-variable` feature in nativescript-6.1.

I hope you can answer a few questions, it might give us some hints to the cause of the problem.

- Are you using the theme extensions from the plugin? On `iOS` this loaded a lot of extra css.
- Do you use some of the other angular features? Like `A11YFontScalePipe`, `GridLayout[a11yRows], GridLayout[a11yColumns]` or `A11yFontScalingObservable`.
- Are you using the latest tns runtime?

**Comment by limux82** [comment]
> Hi @limux82
> 
> Thank you very much for your feedback.
> 
> Our app is still on nativescript 5.4.x, so we haven't run into these issues yet, but of cause they have to be addressed.
> 
> I've just published ~`v6.1.0-alpha.1`~ `v6.1.0-alpha.2` with support for theme-2.0 (including compat-mode) and it should give some performance improvements.
> It uses the new `css-calc` and `css-variable` feature in nativescript-6.1.
> 
> I hope you can answer a few questions, it might give us some hints to the cause of the problem.
> 
> * Are you using the theme extensions from the plugin? On `iOS` this loaded a lot of extra css.
> * Do you use some of the other angular features? Like `A11YFontScalePipe`, `GridLayout[a11yRows], GridLayout[a11yColumns]` or `A11yFontScalingObservable`.
> * Are you using the latest tns runtime?
> * Do you experience the issue both with and without VoiceOver enabled?
> * Are you using AOT?
> * iOS version? Device or simulator?

Basically we are experienced performance problem on all real devices (not with the simulator) above all iPhone 6, 7, 8 (iOS 12.4).
Recently we have migrated the app  from Nativescript 5.4 to 6.0 and then 6.1 (Angular 8). Currently we have in the app the version 1 of Nativescript core theme, but I have tried also with v2. The application is bundled, uglified, but currently we aren't enabled the AOT.

The problems appear both on the normal use of the app and when we activate voice over or enlarge font size. In the app we are using: 
- the fontscale css (~@nota/nativescript-accessibility-ext/scss/fontscales) in order to properly scale the texts in the pages and sometimes change some height/width. We follow the example reported here https://raw.githubusercontent.com/Notalib/nativescript-accessibility-ext/master/src/FontScaling.md
- automation text for the voiceover
- only in one page (we experienced the problem on all pages) we are listening to the FontScaleObservable in order to perform height change in one radListView

I tried the v6.1.0-alpha.2, it seems a little better, but the apps remain slow and I see thousands of errors in the console (**Error: Failed to evaluate css-calc**); the problem I see was reported by you to Nativescript (https://github.com/NativeScript/NativeScript/issues/7786). I tried with the workaround written in the issue, but the performance remains bad
It seems that performance problem is caused when the page is drawn. Using the nativescript profiler I see a lot of CSS recalculation cycles. 

If I removed the import of NotaAccessibilityExtModule, without changes in the page/component/css of the application the app is fast (the route changes are immediate)

**Comment by m-abs** [comment]
Thank you very much for your feedback.

It sounds very much like it is caused by us adding the `a11y-fontscale`-classes to each view in the view-tree.

I've been working on improving the plugin's performance and the 1000s of errors related to `css-calc` should be fixed with `v6.1.0-alpha.4` (it was a very special case in the syntax).
The bug in `tns-core-modules` was fixed in `v6.1.1` so the workaround isn't needed anymore.

The way we added the `a11y-fontscale`-classes triggered multiple updates on all views. The deeper a view were in the view tree, the more times the update were triggered.
The new implementation triggers the update once from the root-view/modal-views and should be much quicker.

Using `~@nota/nativescript-accessibility-ext/scss/fontscales` generates a lot of css-classes.
This slows down UI rendering and we like to remove support for it, probably from `v6.2.0` or `v6.3.0`.

I recommend switching to `css-variables` and `css-calc` instead.

**Comment by martinfrancois** [comment]
We noticed that version `5.4.2` works also for us with NativeScript `6.1.0` and the performance is much better. Is there any reason this would be a bad idea for us to use? Is there any disadvantage to that in comparison to using version 6 of this plugin?

**Comment by m-abs** [comment]
I haven't tried it but I think it will crash on `android`, because of the switch to `androidx` in `nativescript-6.0.0`.

**Comment by m-abs** [comment]
Hi @limux82 

I just had a thought, could you try enabling tracing and giving me the output?

Add this to your `app.module.ts`:
```typescript
import * as trace from 'tns-core-modules/trace';
import { categories } from '@nota/nativescript-accessibility-ext/trace';

trace.setCategories(categories.All.join(','));
trace.enable();
```

This should generate a lot of output, and please check for private strings before posting.

**Comment by limux82** [comment]
Here the trace of the app, I hope can give us some hints

[trace.txt](https://github.com/Notalib/nativescript-accessibility-ext/files/3756990/trace.txt)

**Comment by limux82** [comment]
@m-abs  Do you have any news, suggestion about the performance problem checking the trace log?

**Comment by m-abs** [comment]
Hi @limux82 

Unfortunately the trace didn't help me figuring out what the problem is.

We're in the process of upgrading to nativescript-6.1 and we haven't run into this issue (yet), which makes it very difficult to solve.

The only thing I noticed in your trace is that you use `accessibilityIdentifier`, which we don't use. But it made no difference when I tried doing the same just now.

I'm on a iPhone SE running iOS 12.4.1.

**Comment by limux82** [comment]
Hi @m-abs 

We continue to have poor performance when we add the import of the NotaAccessibilityExtModule with Nativescript 6.4, iOS 13, angular 8
Basically on a RadListView with 10 elements visibile, with font-scale =1 we have an improvement of 30-35% in the rendering of the page only removing the import of NotaAccessibilityExtModule. 

So since for us is a requirement to have the app accessible, I was thinking if there is a way to drive the activation of the plugin (and of the observers / listeners inside) only in determinate conditions at the startup of the app (e.g. font scale > 1, voiceOver active ...) in order to have in cases, where accessibility is not necessary, the same performance of the app without the plugin
Is it possible for you? Is there a way to do that, or has to be implemented?

**Comment by m-abs** [comment]
Hi @limux82 

I'm sorry you're still having this problem.

We don't experience the same with our app, so I'm wondering what differences could cause the problem.

It would require some work to do as you suggest, it should be possible. But I doubt the problem is the observers / listeners.

One thing you could try it using the plugin without the angular module.
Just import `@nota/nativescript-accessibility-ext` (note without the angular subdir) in you `app.module.ts`.

It might be easier to solve the problem, if you contact me on NativeScript's slack, I go by the same nickname.

I'm thinking of adding some profiling to the plugin, when we could try using the `timeline` profile.

**Comment by m-abs** [comment]
Hi @limux82, 

I've just published `v6.4.1` with the changes you've tested.
I'm closing this issue.

Thank you very much for helping us track this down.
